galaxy/api/internal/views/combined.py

```diff
@@ -0,0 +1,101 @@
+# (c) 2012-2019, Ansible by Red Hat
+#
+# This file is part of Ansible Galaxy
+#
+# Ansible Galaxy is free software: you can redistribute it and/or modify
+# it under the terms of the Apache License as published by
+# the Apache Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#
+# Ansible Galaxy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# Apache License for more details.
+#
+# You should have received a copy of the Apache License
+from rest_framework import views
+from galaxy.main import models
+from rest_framework import response
+from galaxy.api.internal import serializers as internal_serializers
+from galaxy.api import serializers as v1_serializers
+
+
+class CombinedList(views.APIView):
+    def get(self, request):
+        # params: page, page size, namespace, name, description, type
+        # order by: name, downloads
+        namespace = request.GET['namespace']
+        page = int(request.GET.get('page', 1))
+        page_size = int(request.GET.get('page_size', 10))
+        type = request.GET.get('type', None)
+
+        collection_filter = {
+            'namespace__name': namespace,
+        }
+
+        repo_filter = {
+            'provider_namespace__namespace__name': namespace
+        }
+
+        if 'name' in request.GET:
+            collection_filter['name__icontains'] = request.GET['name']
+            repo_filter['name__icontains'] = request.GET['name']
+
+        if 'description' in request.GET:
+            collection_filter[
+                'latest_version__metadata__description__icontaines'
+            ] = request.GET['description']
+            repo_filter['description'] = request.GET['description']
+
+        # Avoid loading models if the type is set.
+        if type == 'collection':
+            repos = []
+            repo_count = 0
+        else:
+            repos = models.Repository.objects.filter(**repo_filter)
+            repo_count = repos.count()
+
+        if type == 'repository':
+            collections = []
+            collection_count = 0
+        else:
+            collections = models.Collection.objects.filter(**collection_filter)
+            collection_count = collections.count()
+
+        start = (page * page_size) - page_size
+        end = page * page_size
+
+        collection_results = collections[start:end]
+
+        # If the collections fill the whole page, don't bother loading the
+        # repos
+        if (len(collection_results) == page_size):
+            repo_results = []
+        else:
+            # Django gets cranky if you try to slice a queryset with a negative
+            # index
+            r_start = start - collection_count
+            r_start = 0 if r_start < 0 else r_start
+
+            repo_results = repos[r_start:end-collection_count]
```

Comment:
I think you actually want `len(collection_results)` rather than `collection_count`, because you want to offset the repository results by the number of collections in this page, not the total number of collections.

Reply:
I think my approach is correct because in this case `end` is the index of the last item in the list if the list was a combined list of repos and collections so we need to offset the starting point for repos by the total number of collections

Lets say we have 21 collections and 40 repos. On page 3 (with page size 10) `end=30` and  we would expect to see `collection_results=collections[20:21]` and `repo_results=repos[0:9]`. If we used `len(collection_results)` then the function would return `repos[0:30-1]`, which doesn't work.